Article about architecture review processes (#2)
steffenmueller4/steffenmueller4.github.io: _posts/2021-05-16-overview-of-architecture-review-processes.md

> ---

Reviewer: General feedback:

1. The first glance is a wall of text. One idea to brighten it up would be to create/add a few illustrations that underline or support your text. Two or three might be full enough.
2. What do you think about a 3 sentence "tl;dr" at the beginning of the article?
3. The article has multiple long sentences. Small/short sentences are easier to read and understand. Maybe you want to consider ease out a few sentences.
4. The reading flow is getting easier. The first part is hard to read. The middle section is getting easier. And the design doc part is a good flow. Not sure what the root cause is. One observations: In the design doc part, the quotes are embedded into the reading flow/in the story.
5. I have the feeling you are using a lot of filler words. Some words/parts can be removed without removing the meaning/context. This would make the text crisper.
6. I love that you provide decent references. Thanks for doing so. Very rare.

For point 3, 4 and 5, tools like https://hemingwayapp.com/ and Grammarly can help.

Overall a pretty solid article.
A little bit hard/long to read, but while tackling 4, 5, and 6, it is getting shorter.

What I would love to see:
1. A direct comparision for those three (maybe in a table?)
2. Your personal experience and opinion when to use what
3. One example aerchitecture and applied every process on it

I am pretty sure those requests should not go into this article.
They might be food for thought for future articles.

Good work, Steffen!

Author: Added 3 figures about the processes. If you could take a look, would be very cool ❤️

Author: I also added a basic table to summarize the three different process in the end.

layout: post
date:   2021-05-29 09:02:08 +0100
title: "Overview of different Architecture Review Processes"
categories:
  - Architecture
  - Organization
  - Architecture Review Process
published: true
---
Recently, I have been asked about what software architecture review processes exist in order to better steer architecture.
An architecture review aims at different goals such as finding software design issues early in the development before they get costly.

> Architecture review processes, for example, formalize different steps.

Author: I do not think so, because the first sentence is about architecture review, the second is about architecture review processes. You can also perform an isolated architecture review without a formalized process such as a one-time architecture review from an external company. However, establishing a process is something different.

They, furthermore, define input and output to/from architecture reviews.
This article describes the different architecture review process approaches.

## What are Architecture Reviews and Architecture Review Processes?

An architecture review, in a nutshell, is a mechanism for increasing the likelihood that a software/system architecture will be complete, consistent, and, thus, good {% cite Maranzano2005 %}.
A good software architecture is important, otherwise the development can become slower.
It can get more expensive to add new capabilities in the future.
High internal quality, on the other side, can lead to faster delivery of new features; simply, because there is less cruft to get in the way {% cite Fowler2019 %}.

Essentially, architecture reviews aim at {% cite Maranzano2005 %}:
  1. finding software design problems early in development before they get costly.
  1. building projects based on best practices and transfer this knowledge across the organization.
  1. improving the organization's software quality and operations (-> documentation).

> Therefore, architecture review processes formalize different steps, the parties involved, or input and output to architecture reviews.

Author: Mmh... may be, but I do not have any clue to improve this sentence, then...

Over the last decades, different architecture review processes have emerged.

As every such process can be modified and adapted to the unique requirements of a specific company, we will focus on three different prototypical processes that are described in literature:
  1. The "Classical" Architecture Review process explained by Maranzano et al. in {% cite Maranzano2005 %}.
  1. The Architecture Decision Records initially described by Nygard in {% cite Nygard2011 %}.
  1. The Lightweight Request for Comment (RFC)/Design Document approach explained in {% cite Ubl2020 %}, {% cite Winters2020 %}, {% cite Orosz2020 %}, {% cite Orosz2021 %}, {% cite Zimmermann2019 %}, {% cite Gonchar2020 %}, and {% cite Mozilla2020 %} from the perspectives of different organizations - sometimes called Lightweight RFC and sometimes Design Document approach. The approach has also been mentioned in the ThoughWorks Tech Radar Vol. 24 trial area as Lightweight RFC approach in {% cite ThoughtWorks2021 %}.

In the next sections, we will describe these three different prototypical architecture review processes.

> ## The "Classical" Architecture Review Process

Reviewer: In comparison to the "Architecture Decision Records" chapter, this is kinda hard to read.
An info graphic will be 100% beneficial.

The "Classical" Architecture Review process is described by Maranzano et al. in a paper based on common architecture review processes at AT&T, Avaya, Lucent, and Millenium Services {% cite Maranzano2005 %}.
The paper stems from 2005, the idea is based on AT&T's practices from the late 1980s.
A lot of established companies (still) use this approach in some variation.

### Parties

The approach considers three primary parties in an architecture review {% cite Maranzano2005 %}:
  1. The project team which requests the architecture review.
  1. The review team which consists of experts assembled for the review on the basis of their expertise, their independence from the specific project, and their ability to conduct themselves appropriately in a potentially difficult interpersonal situation.
  1. The Architecture Review Board (ARB) which is a standing team that oversees the review process and its effect on the organization.

> Moreover, there are further roles in the overall process - depending on the size of the enterprise and if these roles are required:

Author: You are very right. I will limit myself to "organization" here, because it does not always have to be a company or enterprise. It can also be a part of a company or an open source community.

### Process

The overall review process bases on a few general principles such as that a clearly defined problem statement as the basis of the architecture, reviews are open processes, and companies conduct reviews for the project’s benefit (see also: {% cite Maranzano2005 %} for the full list of five principles).

The review process follows roughly four phases.
The specific implementation may vary.
  1. Screening Phase: The project team requests a review. The project staff and ARB review the request to determine whether a review would benefit the project. If a review should be conducted, the ARB selects a review angel (see also: previous section) to oversee the project’s review process.
  1. Preparation Phase: The ARB selects a review team and a review leader. Togehter with the project team, they determine the number of reviews and the initial review’s date and agenda. The review team, project team, review leader, and review angel verify that the project that should be reviewed has a clearly defined problem statement driving the architecture and an appropriate documentation as a basis for the review.

>   1. Review Meeting Phase: During the review, the members of the project present the mentioned problem statement and how the proposed architecture solves it. The reviewers ask further questions in order to figure out issues they believe could make the project an incomplete or inadequate solution to the problem. Finally, the review team meets privately and generates a report. The report is presented to the project team. Furthermore, management alerts can be raised, etc.

Reviewer: > Furthermore, management alerts can be raised, etc.

For me, it is hard to think about what can be other things (behind the etc.).
Either you provide more examples or you drop this etc.

  1. Follow-Up Phase: The review team delivers the report with the review’s findings to the project team within certain time frame such as 15 days after the review meetings are done. The project team (and, if a management alert has been raised, the management) must respond within two weeks.

> For more details, we refer to {% cite Maranzano2005 %}.

Reviewer: ```suggestion
```

I would drop this.
Why? Many information were provided + if someone want to go deeper, the person has multiple options above.

### Artifacts

> As artifacts, Maranzano et al. mention three different types of artifacts {% cite Maranzano2005 %}:

Reviewer: ```suggestion
Three different types of artifacts are getting created:
```

Sure, it was Maranzano et al. who defined it. However, it is your article ;)

Another thought: I am reading this article mainly because of the topic. Not because of the Author (Maranzano).

  1. There are checklists for architects to prepare for the review and checklists for the reviewers to do the review and figure out more of the architecture. The checklists serve as a collection of organizational knowledge and should be maintained.

>   1. There is input to the review. This can be documentation of the system requirements, functional requirements, or informal documentation. Maranzano et al. mention specifically that the architecture should be designed based on a clear problem statement tackling the functional and qualitative requirements, the costs, and the timeline. A document describing these points may be a specific design document or something like that.

Reviewer: ```suggestion
  1. There is input to the review. This can be documentation of the system requirements, functional requirements, or informal documentation. The architecture should be designed based on a clear problem statement tackling the functional and qualitative requirements, the costs, and the timeline.
```

> A document describing these points may be a specific design document or something like that.

This is the first time in the classical review that you mention design document.
At this point, I don't know what this is.
For me, it reads that the requested artifact has no formatting rules. Hence, I would drop it.

>   1. There is output from the review such as the review report, an optional management alert letter, an optional set of issues, etc.

Reviewer: > an optional set of issues

Issues == Problems?
Or Issues == Tickets in a ticket system?

Author: That's not so clear. May be bigger problems (= challenges) or simple tasks that the project team has to work on. I will enhance the sentence a little bit to make this a little bit clearer to the reader.

### Conclusion

The "Classical" Architecture Review process is a well-established and formalized process.
It is (still) widely used.
Especially, bigger companies or, at least, companies with dedicated architects often make use of a process explained by Maranzano et al. in some variation.

On the one hand, the formalized process as well as the clear roles within the process help to achieve the goals of architecture reviews.

> Also, there are different options to vary the process in order to, for example, have less parties involved or reduce the overhead.

Reviewer: > reduce the overhead

Which overhead? This is the first time this is mentioned. Up to know this was a described process.
It seems that you are mixing here the original process description by the original author with your personal opinion.

Author: You are right. Will remove the "overhead".

> In some companies, there is, for instance, no ARB or review team but an architecture guild which is performing architecture reviews instead.

Reviewer: Maybe linking the word "guild" to an explanation (e.g. community of practice). Why? Not everyone might know what this is.

> Another option is to reduce the overhead of the artifacts and, therefore, only give a presentation about the architecture to be reviewed - however, I personally think that there should be a written architecture/design document.

Reviewer: ```suggestion
Another option is to reduce the overhead of the artifacts. Only give a presentation about the architecture to be reviewed - however, I personally think that there should be a written architecture/design document.
```

On the other hand, the formalized process and the overall overhead for the "Classical" Architecture Review process is also often the point to be critizied mostly.
The overall process makes a lot of overhead for smaller projects or architecture changes which have to be reviewed.
In the paper, Maranzano et al. mention that the review preparation time may take up to 6 weeks, the review itself between 1 and 5 days.

> So, the overhead can be substantial to a company in the prototypical form.

Reviewer: 3 times the word _overhead_. Maybe rephrasing it a bit.

Reviewer: > prototypical form

a startup?

In sum, the entire prototypical approach does not appear to be very agile, but there are also ways to improve the overall process.
Nevertheless, it has the potential to balance architectural as well as company-political aspects.

## Architecture Decision Records

Architecture Decision Records (ADR) were initially proposed by Nygard in 2011 in {% cite Nygard2011 %}.

> Nygard stated that architecture in an agile context has to be described and defined differently.

Reviewer: differently. differently than what? In comparison to the classical approach?

Therefore, Nygard proposes ADR as small documents which concentrate on tracking the motivation, rationale, and consequences behind certain architectural decisions made during different points in time in the project in a very structured and agile way.
"We will keep a collection of records for 'architecturally significant' decisions: those that affect the structure, non-functional characteristics, dependencies, interfaces, or construction techniques." (Nygard, 2011)
Essentially, the small and structured ADR help to find out about motivation, rationale, and consequences of previous decisions at the time being written and reviewed as well as later on.
The ADR form a "decision log" {% cite ADR2021 %}.

For more information, we refer to [https://adr.github.io/](https://adr.github.io/).

### Parties

When working with ADR, we can distiguish the following parties:
  1. The project team/creator(s) of an ADR,
  1. The ADR reviewer(s)

### Process

When there is an architecturally significant decision to be made, the project team, the involved architect, or - in general - persons who are involved in the decision making create an ADR.

> Based on Nygard's proposal, ADR typically contain 5 sections {% cite Nygard2011 %}:

Reviewer: ```suggestion
An ADR typically contains 5 sections {% cite Nygard2011 %}:
```

ADR can be stored in GIT or in a wiki.
There are tools for managing ADR in GIT, e.g., [https://github.com/npryce/adr-tools](https://github.com/npryce/adr-tools).
In my personal opinion, it is recommendable to store ADR in a public location such as a wiki for better accessibility from an potentially interested audience.
However, the tool should be able to track changes to the ADR in a history to identify authors and revert changes if needed.
Moreover, a comment functionality for reviewing an ADR is beneficial.

Every ADR and, thus, every decision is an own small document (see also: "Decision log" in the previous section).
ADR can get an ID and be referenced in other documentation or wiki pages.
New ADR can change aspects of previous ADR.
If that happens, they can change the status of the previous ADR to "Superseded".

### Artifacts

The ADR itself is the only artifact in this prototypical architecture review process.
Reviews on the ADR should be done on the ADR itself when the ADR is created.

In recent years, different templates have appeared for ADR.
In {% cite Henderson2021 %}, Henderson gathered different ADR example templates.

> However, ADR can be combined with other ways to document architectures such as Design Documents (see also: Design Document or Lightweight Request for Comment Approach):

Reviewer: > see also: Design Document or Lightweight Request for Comment Approach)

We are on the web. Maybe adding an anchor link? ;)

Author: I tried to set up anchors. Was not successful so far. But I will reconsider that :-)

### Conclusion

ADR are well-suited and widely accepted for documenting architecture decisions {% cite Richards2020 %}.
The decision log formed by ADR can be a very essential part to understand the history and the current state of an architecture.
Based on the history and the current state, you can make better decisions in the future.
ADR and their proposed templates condense architecture decisions to the essential parts.
Specifically, when they are stored publicly accessible as a basis for discussions, they are a very good tool for architecture work.

ADR and their strict templates are limited when it comes to the bigger picture of changes.
A good ADR should be written neatly and to the point.
This, however, may lead to important aspects of changes not being properly described in an ADR.
"The ADR is a snapshot of information uncovered during the creation process." {% cite Kuenzli2019 %}
Kuenzli suggests in {% cite Kuenzli2019 %} that also other types of documents should be created in an architecture design process in addition to the ADR.
For example, Design Documents can complement ADR.

> Design Documents are more verbose and capture more information gathered during research phases.

Reviewer: In this chapter, you conclude the ADR.
These two lines sound rather more like an ADR <-> Design Document comparison.

Author: Mmh, yes, somehow you are right. But I have no idea how to improve this part. Essentially, ADR are often not enough documentation needed. At OBI, we also tried to do both for a bigger project: DD + ADR being closer described in some parts of the DD.

The ADR can be created as the outcome of the overall architecture design process.

## Lightweight Request for Comment/Design Document Approach

> Another approach is the Lightweight Request for Comment (RFC)/Design Document (DD) approach.

Reviewer: ```suggestion
Another approach is Lightweight Request for Comments (RFC) or Design Documents (DD).
```

> Essentially, this approach is about writing a DD as the review artifact, sharing this document across the organization in some way, and discussing as well as improving the DD - and the system in design - together with the reviewers.

Reviewer: ```suggestion
Essentially, this approach is about writing a DD as the review artifact, sharing this document across the organization, and discussing as well as improving the DD - and the system in design - together with the reviewers.

"in some way" doesn't add any value.
```

It is a more informal way of documenting software architecture at a certain point in time {% cite Ubl2020 %} - originally, considered for the design phase but can be applied to any rework of architectural aspects (solution idea) before the actual (code) work is done.

> The Lightweight RFC/DD approach is called differently by various authors - sometimes called Lightweight RFC (or only RFC) and sometimes DD approach.

Reviewer: ```suggestion
The Lightweight RFC/DD approach is called differently by various authors - sometimes Lightweight RFC (or only RFC) and sometimes DD.
```

Author: I changed the entire paragraph. Hopefully, to the better ;-)

Orosz additionally states that the "[...] type of information pushed to people in an organization shapes the culture considerably. [...] [It] sets a tone of trust and responsibility." {%cite Orosz2020 %}

On top of that community thinking, the approach considers the power of the written word.
"Writing and sharing that writing with others creates accountability. It also almost always leads to more thorough decisions." {% cite Orosz2020 %}

### Parties

The DD should be written by the team actually working on the solution idea during the work on finding the solution idea, but before implementating the solution (see also: {% cite Ubl2020 %} and {% cite Orosz2021 %}).

As soon as the solution idea stabilizes, the DD is shared with the entire company - or at least with a huge amount of interested people.
The interested people should review and discuss the DD and the solution idea to improve it in an company-open and lightweight process.

### Process

Ubl {% cite Ubl2020 %} and Orosz {% cite Orosz2021 %} suggest roughly the following process (see: {% cite Ubl2020 %} and {% cite Orosz2021 %}):

>   1. The team starts with the business problem and brainstorms solution ideas. During that phase, the DD shoud be started and iterated rapidly within the team itself until the solution idea/DD stabilizes.

Reviewer: ```suggestion
  1. The team starts with the business problem and brainstorms solution ideas. During that phase, the DD should be started and iterated rapidly within the team itself until the solution idea/DD stabilizes.
```

Main goal of the DD is the communication aspect: A DD should identify issues of a solution idea as early as possible in the project lifecycle, help to achieve consensus around a design in the organization, ensure the consideration of cross-cutting concerns, distribute knowledge about solution ideas, and document the solution.
In order to achieve that, a certain structure has emerged focusing on architecturally important aspects of a solution idea at Google (see, e.g.: {% cite Ubl2020 %}):
  * Context and Scope Goals and Non-goals
  * The actual Design
    * System-context Diagram
    * APIs
    * Data Storage
    * etc.
  * Alternatives Considered
  * Cross-cutting Concerns

Thereby, the template is not fixed and sections are optional depending on the actual solution idea.
Also, solution ideas may require other aspects to be described.
In general, the DD should focus on the discussion of trade-offs of the solution ideas that were considered during decisions.

"Design docs should be sufficiently detailed but short enough to actually be read by busy people." {% cite Ubl2020 %}
Recommendations reach from 1 up to 20 pages depending on the problem and solution idea.

> As writing a DD is overhead, the decision whether to write a DD comes down to the core trade-off of deciding whether a review is beneficial.

Reviewer: I love this one. Pure <3

An example DD is available [here](https://www.industrialempathy.com/posts/design-doc-a-design-doc/).

### Conclusion

The Lightweight RFC/DD approach is a good way to let the entire company participate in and improve the design of the company's systems.
Examples from big software companies such as Google show that this approach works well and can produce good system designs while spreading the knowledge in the company.

However, the approach requires active participation of the company's developers or architect's.
Not every company is ready for such an approach - this may also originate from a different working style or culture such as a very hierarchical organization or a company with a strong and well-established architecture team that does not want to involve other parties.

Another challenge for the Lightweight RFC/DD approach is the question about when a DD should be created and how much work should be put into a DD.
Here, a company needs to allow teams to spend time in writing DD and gathering the feedback.
If that is not possible, the entire approach may not work well.

> Adaptations to the process may prescribe the when to write a DD and how much time for feedback is allowed.

Reviewer: Good points. Kudos.
